### ISSUE TITLE
Creating Mapbox tileset from go database

### ISSUE THREAD

**Issue by geohacker** [IFRCGo/go-frontend#1474]
With the work we have been doing in https://github.com/IFRCGo/go-api/pull/815 we will soon have all geometries for countries and districts in the Go database. The goal is to have go database become the canonical source of geo data for go-api and front-end.

This means that we can allow admins to edit names, fix centroid, adjust bbox, overwrite geometries etc from within the go-api Django Admin. To make sure these changes are reflected on the front-end maps, we need a workflow to generate Mapbox tilesets from the database and push them to Mapbox so it updates all maps.

What we should do:
* Write a management command that creates a mapbox vector tileset using [Tippecanoe](https://github.com/mapbox/tippecanoe) 
* Upload this to the an existing or versioned map id
* Update style as necessary.

We can only do this once all geometries are imported and inconsistencies like #1456, #1473 and #1464 are resolved first in the shapefile, then imported to the database. In the meantime, we will work on the pipeline while the data is all getting ready perhaps by mid-late August.

cc @batpad

**Comment by batpad** [comment]
Moving to 4.4.

**Comment by geohacker** [comment]
We decided to use the [Mapbox Tiling Service](https://docs.mapbox.com/mapbox-tiling-service/) for this purpose. I've built out the workflow in this PR https://github.com/IFRCGo/go-api/pull/933. Country tileset is processed as expected, but districts are failing. 

Unfortunately there's no error or logs from the Mapbox side that's visible to us using the [tilesets-cli](https://github.com/mapbox/tilesets-cli). When I run the status command, this is what comes back:

```
root@b36819f481d0:~/go-api# tilesets status go-ifrc.go-districts
{"id": "go-ifrc.go-districts", "latest_job": "cki4j413u001207ma8cmo0dy5", "status": "failed"}
```

This is after many hours (> 10 hours) 
I ran tippecanoe manually on my computer with the same dataset and was able to generate an mbtiles file successfully. 


@jhenshall @batpad I plan to email Mikel to see if he can help us get some more information about this.

**Comment by geohacker** [comment]
We resolved the MTS error by reducing the coordinate precision and setting a maxzoom of 10 based on Mapbox's suggestion. MTS does fail occasionally with little helpful errors but the script should work and when run as a management command will update the tilesets.

I think all items in this ticket are done.

**Comment by geohacker** [comment]
Tileset generation code is now in staging. Once this is merged to production, we will import and update all data changes in the database and then run the management command to update the tileset.

**Comment by jhenshall** [comment]
Hey @geohacker - can we change the zoom levels for the tilesets, especially districts as you can't see them until zoomed too far in. Not certain what best here. The disputed boundaries uploaded direct to Mapbox are z0-8 (and cannot be changed unless we use API). Tempted to make all layers the same so they continue to line up (more or less) as you zoom closer? Could be overkill though.

**Comment by geohacker** [comment]
@jhenshall yes! I highly recommend to make sure we pick a zoom level that's needed and not go fully overboard. It introduces more friction points and higher chances of MTS failure and a Mapbox bill. Can you tell me what zoom levels are needed?

**Comment by jhenshall** [comment]
Proposing the zoom level and naming for the MTS layers:

- GO_Countries_v1 - z0-8
- GO_Districts_v1 - z3-8
- GO_Country_Centroids_v1 - z0-4
- GO_District_Centroids_v1 - z3-6

**Comment by geohacker** [comment]
@jhenshall This is now the zoom levels for the tilesets. They are a bit bigger range than listed above:

Countries minzoom  0, maxzoom  10
Districts minzoom 3, maxzoom 10
Countries centroids minzoom 0, maxzoom 10
Districts centroids minzoom 3, maxzoom 10

https://github.com/IFRCGo/go-api/pull/1028

**Comment by jhenshall** [comment]
When building the larger districts tileset, the VM with 4GB RAM was insufficient and crashed. This is now a blocker to finishing the Mapbox build. @batpad @geohacker - I hope this captures the issue but perhaps we need a new ticket for the VM issue?

**Comment by jhenshall** [comment]
Started building the styles in Mapbox with the new tilesets. A few issues to note as discussed @batpad @geohacker:

- Nepal districts missing
- Morocco and Serbia missing (we updated the geometries)
- Add country ISO (and maybe name) to district attributes - currently used to focus on countries of interest on emergency page
- Add disputed attribute to countries (we need it to style them)

**Comment by geohacker** [comment]
@jhenshall I can do 3, 4, 5 tomorrow. @batpad let me know if we can coordinate with Zoltan on checking the nepal and other geometries are actually missing the database.

**Comment by szabozoltan69** [comment]
Nepal is in the country table:
```
... api_country c where cd.country_id = c.id and c.record_type=1 and name like '%epal%'
 country_id | name  | iso 
------------+-------+-----
        123 | Nepal | NP
```

**Comment by geohacker** [comment]
Here are things to do to wrap this up:
1. Merge https://github.com/IFRCGo/go-api/pull/1103 and deploy
2. Update serbia and morocco geometry projection *SHOULD BE TESTED ON STAGING FIRST*:
```
ALTER TABLE api_countrygeoms SET geom=ST_Transform(ST_SetSRID(geom, 3857), 4326) WHERE country_id=271;
ALTER TABLE api_countrygeoms SET geom=ST_Transform(ST_SetSRID(geom, 3857), 4326) WHERE country_id=119;
```
3. Update nepal geometries with the [new shapefile] (https://drive.google.com/file/d/1LjNDbBuAJ3kDldwgbIb6CV0x5aEo1QE9/view?usp=sharing). By running `python manage.py import-admin1-data 20210610_NepalRC_Nepal-adm1_CODE-UPDATE.shp --update-geom --update-centroid --update-bbox` *SHOULD BE TESTED ON STAGING FIRST*

**Comment by szabozoltan69** [comment]
Did you mean "UPDATE api_countrygeoms..." in the 2nd point?

**Comment by geohacker** [comment]
@szabozoltan69 indeed! I meant `UPDATE` I just fixed the query above

**Comment by szabozoltan69** [comment]
The two updates are run successfully on staging:
```
postgres=> UPDATE api_countrygeoms SET geom=ST_Transform(ST_SetSRID(geom, 3857), 4326) WHERE country_id=271;
UPDATE 1
postgres=> UPDATE api_countrygeoms SET geom=ST_Transform(ST_SetSRID(geom, 3857), 4326) WHERE country_id=119;
UPDATE 1
```

**Comment by geohacker** [comment]
@szabozoltan69 excellent! can you paste the output of the following query in a txt file and share? 

`SELECT ST_AsGeoJSON(geom) as geom FROM api_countrygeoms WHERE country_id=271`

**Comment by szabozoltan69** [comment]
I've run it, but to get the file from the remote environment needs some effort...

**Comment by szabozoltan69** [comment]
Shall I run this?
```
python manage.py import-admin1-data 20210610_NepalRC_Nepal-adm1_CODE-UPDATE.shp --update-geom --update-centroid --update-bbox
```

**Comment by szabozoltan69** [comment]
@geohacker – So 1, 2, 3 steps are ready on staging.

**Comment by geohacker** [comment]
@szabozoltan69 ⚡  fast! could you run a query fetch all the nepal districts from `api_districtgeoms` on staging and share here so @jhenshall can confirm it looks good.

**Comment by szabozoltan69** [comment]
[NepalDistricts.zip](https://github.com/IFRCGo/go-frontend/files/6630825/NepalDistricts.zip)
This is the result of the below query:
```
select id, name from api_country where name like '%epal%';
 id  | name  
-----+-------
 123 | Nepal  (to check id)

select d.name, geom from api_districtgeoms g join api_district d on(g.district_id=d.id and country_id=123);

**Comment by geohacker** [comment]
Looks like the `UPDATE` to fix projection didn't work as expected. I will do some more experiments and confirm when this is ready for production

**Comment by jhenshall** [comment]
Very happy that this is now DONE! Thanks to @geohacker @batpad and @vdeak!! This is a satisfying ticket to close. 😄
